File: Data Structures & Algorithms/decode-ways/submission-1.cpp
class Solution {
public:
    int solve(int index,string& s,auto& dp){
        if(index >= s.size())return 1;
        if(dp[index] != -1)return dp[index];
        // single char
        int ans = 0;
        if(s[index] >= '1' && s[index] <= '9'){
            ans = solve(index + 1, s, dp);
            if(index < s.size()-1){
                if(s[index] == '1' && s[index+1] >= '0' && s[index+1] <= '9'){
                    ans += solve(index+2,s, dp);
                }
                if(s[index] == '2' && s[index+1] >= '0' && s[index+1] <= '6'){
                    ans+= solve(index+2,s ,dp);
                }
            }
        }
        return dp[index] = ans;
    }   
    int numDecodings(string s) {
        vector<int> dp(s.size()+2,0);
        dp[s.size()] = 1;
        dp[s.size()+1] = 1;
        for(int index = s.size()-1; index >= 0 ; index--){
            int ans = 0;
            if(s[index] >= '1' && s[index] <= '9'){
                ans = dp[index + 1];
                if(index < s.size()-1){
                    if(s[index] == '1' && s[index+1] >= '0' && s[index+1] <= '9'){
                        ans += dp[index+2];
                    }
                    if(s[index] == '2' && s[index+1] >= '0' && s[index+1] <= '6'){
                        ans+= dp[index+2];
                    }
                }
            }
            dp[index] = ans;
        }
        return dp[0];
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.java
class Solution {
    public void solve(int[] nums,int target,List<List<Integer>> ans,List<Integer> temp,int index){
        if(target == 0){
            ans.add(new ArrayList<>(temp));
            return;
        }
        for(int i = index; i < nums.length; i++){
            if(i > index && nums[i] == nums[i-1])continue;
            if(nums[i] <= target){
                temp.add(nums[i]);
                solve(nums,target-nums[i],ans,temp,i+1);
                temp.remove(temp.size()-1);
            }
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> temp = new ArrayList<>();
        solve(candidates,target,ans,temp,0);
        return ans;
    }
}
